File: backend/src/main/java/com/runninggu/server/contest/domain/Contest.java
package com.runninggu.server.contest.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalTime;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

@Entity
@Table(name = "contest")
public class Contest {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "canonical_key", nullable = false, unique = true, length = 255)
    private String canonicalKey;

    @Column(nullable = false, length = 255)
    private String name;

    @Column(nullable = false, length = 32)
    private String region;

    @Column(nullable = false, length = 255)
    private String place;

    @Column(name = "road_address", length = 255)
    private String roadAddress;

    @Column(precision = 10, scale = 7)
    private BigDecimal lat;

    @Column(precision = 10, scale = 7)
    private BigDecimal lng;

    @Column(name = "contest_date", nullable = false)
    private LocalDate contestDate;

    /** 대회 시작 시각은 timezone 변환을 하지 않는 KST 벽시계 값이다. (SPEC §6.6) */
    @JdbcTypeCode(SqlTypes.LOCAL_TIME)
    @Column(name = "start_time")
    private LocalTime startTime;

    @Enumerated(EnumType.STRING)
    @Column(name = "source_status", length = 32)
    private ContestRegistrationStatus sourceStatus;

    @Column(name = "apply_start")
    private LocalDate applyStart;

    @Column(name = "apply_end")
    private LocalDate applyEnd;

    @Column(length = 255)
    private String organizer;

    @Column(name = "official_url", length = 2048)
    private String officialUrl;

    @Column(name = "detail_url", length = 2048)
    private String detailUrl;

    @Column(name = "image_url", length = 2048)
    private String imageUrl;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 32)
    private ContestCategory category;

    @Column(nullable = false)
    private boolean active;

    @Column(name = "checked_at", nullable = false)
    private Instant checkedAt;

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    protected Contest() {}

    public static Contest create(String canonicalKey, Instant updatedAt) {
        Contest contest = new Contest();
        contest.canonicalKey = canonicalKey;
        contest.active = true;
        contest.updatedAt = updatedAt;
        return contest;
    }

    /** 서버용 snapshot의 canonical 값을 전부 교체한다. (SPEC §8.2, 결정-40·47) */
    public void update(
            String canonicalKey,
            String name,
            String region,
            String place,
            String roadAddress,
            BigDecimal lat,
            BigDecimal lng,
            LocalDate contestDate,
            LocalTime startTime,
            ContestRegistrationStatus sourceStatus,
            LocalDate applyStart,
            LocalDate applyEnd,
            String organizer,
            String officialUrl,
            String detailUrl,
            String imageUrl,
            ContestCategory category,
            Instant checkedAt,
            Instant updatedAt) {
        this.canonicalKey = canonicalKey;
        this.name = name;
        this.region = region;
        this.place = place;
        this.roadAddress = roadAddress;
        this.lat = lat;
        this.lng = lng;
        this.contestDate = contestDate;
        this.startTime = startTime;
        this.sourceStatus = sourceStatus;
        this.applyStart = applyStart;
        this.applyEnd = applyEnd;
        this.organizer = organizer;
        this.officialUrl = officialUrl;
        this.detailUrl = detailUrl;
        this.imageUrl = imageUrl;
        this.category = category;
        this.checkedAt = checkedAt;
        this.updatedAt = updatedAt;
    }

    public void updateActive(boolean active, Instant updatedAt) {
        if (this.active != active) {
            this.active = active;
            this.updatedAt = updatedAt;
        }
    }

    public Long getId() {
        return id;
    }

    public String getCanonicalKey() {
        return canonicalKey;
    }

    public String getName() {
        return name;
    }

    public String getRegion() {
        return region;
    }

    public String getPlace() {
        return place;
    }

    public String getRoadAddress() {
        return roadAddress;
    }

    public BigDecimal getLat() {
        return lat;
    }

    public BigDecimal getLng() {
        return lng;
    }

    public LocalDate getContestDate() {
        return contestDate;
    }

    public LocalTime getStartTime() {
        return startTime;
    }

    public ContestRegistrationStatus getSourceStatus() {
        return sourceStatus;
    }

    public LocalDate getApplyStart() {
        return applyStart;
    }

    public LocalDate getApplyEnd() {
        return applyEnd;
    }

    public String getOrganizer() {
        return organizer;
    }

    public String getOfficialUrl() {
        return officialUrl;
    }

    public String getDetailUrl() {
        return detailUrl;
    }

    public String getImageUrl() {
        return imageUrl;
    }

    public boolean isActive() {
        return active;
    }

    public Instant getCheckedAt() {
        return checkedAt;
    }
}

File: android/app/src/main/java/com/runninggu/app/data/model/Contest.kt
package com.runninggu.app.data.model

import com.runninggu.app.domain.EventType
import com.runninggu.app.domain.RegistrationStatus
import java.time.LocalDate
import java.time.LocalTime

/**
 * 대회 하나. (SPEC §6.2 canonical Contest 계약)
 *
 * 서버 응답과 번들 assets 이 **둘 다 이 모델로 들어온다.** 원천마다 JSON 모양이 다르므로
 * DTO 는 따로 두고 매퍼가 여기로 모은다 — 화면과 도메인은 출처를 몰라도 된다.
 *
 * [regStatusFallback] 을 그대로 쓰지 말 것. 표시할 때는 `regStatusOf()` 로 오늘 기준
 * 재계산한다(SPEC §5.5) — 번들과 캐시는 낡기 때문이다.
 */
data class Contest(
    /**
     * 화면·내비게이션용 키. **서버 호출에 쓰지 않는다.**
     *
     * 번들은 크롤 원천의 `externalId`("roadrun-41543"), 서버는 canonical id 를 문자열로 담는다 —
     * 두 값은 표기 차이가 아니라 **다른 namespace** 다.
     */
    val id: String,
    /**
     * canonical `CONTEST.id`. 번들 항목은 **null** 이다. (API 명세 §3-1 · §7-C)
     *
     * 상세·찜·동선 생성처럼 서버를 부르는 동작은 이 값이 있을 때만 한다. 번들 항목은
     * 서버 데이터로 교체되기 전까지 오프라인 표시용이다 — [id] 를 숫자로 바꿔 보내면
     * `NumberFormatException` 이 난다.
     */
    val serverId: Long? = null,
    val name: String,
    val region: String,
    val venue: String,
    val date: LocalDate,
    val startTime: LocalTime?,
    val eventTypes: List<EventType>,
    val regStart: LocalDate?,
    val regEnd: LocalDate?,
    /** 원천이 준 접수 상태. 날짜로 단정할 수 없을 때만 쓴다. */
    val regStatusFallback: RegistrationStatus?,
    val organizer: String?,
    val officialUrl: String?,
    val detailUrl: String?,
    val imageUrl: String?,
    val lat: Double?,
    val lng: Double?,
    val category: String?,
    /** 원본 스냅샷을 확인한 날짜. 카드에 "{출처} · 확인 MM.DD" 로 쓴다(A3). */
    val checked: LocalDate?,
    /**
     * 서비스 중인 대회인가. (결정-46)
     *
     * false 면 화면이 흐리게 그리고 "정보 제공 종료" 를 붙이며, 동선 만들기 CTA 를 막고
     * 인근 축제를 부르지 않는다(§4.6). 번들에는 이 개념이 없어 항상 true 다.
     */
    val active: Boolean = true,
    val sources: List<String>,
) {
    /** 좌표가 없으면 지도·인근 축제·동선 생성을 쓸 수 없다. (§6.2) */
    val hasLocation: Boolean get() = lat != null && lng != null

    /** 서버 상세·찜·동선 생성을 부를 수 있는가. 번들만 있는 대회는 false 다. */
    val isServerBacked: Boolean get() = serverId != null
}
